### PR TITLE
Solutions for Lecture 3

### PR DESCRIPTION
**Pull request by fetsh**
### Solutions for Lecture 3

cc @chshersh @vrom911

### CODE REVIEW

File: src/Lecture3.hs [hunk 1]
```diff
@@ -79,7 +82,10 @@ Tuesday
   would work for **any** enumeration type in Haskell (e.g. 'Bool',
   'Ordering') and not just 'Weekday'?
 -}
-next = error "TODO"
+next :: (Eq a, Bounded a, Enum a) => a -> a
+next el = if el == maxBound
+          then minBound
+          else succ el
```
Comment (chshersh) [inline src/Lecture3.hs:88]: Bonus points for solving all extra challenges 💯 🔝

File: src/Lecture3.hs [hunk 2]
```diff
@@ -89,8 +95,23 @@ weekday to the second.
 >>> daysTo Friday Wednesday
 5
 -}
-daysTo = error "TODO"
 
+-- Using this ["trick"](https://t.me/c/1573945457/940) with `forall`
+-- or we could just use Weekday in types as we already have "days" in
+-- function name or even put `7` in place for len in 
+-- otherwise = len + (from2 - from1) :)
```
Comment (chshersh) [inline src/Lecture3.hs]: I would specify the type of this function to work explicitly with `Weekday`. Indeed, it enabled some optimizations like being able to use the constant `7` directly. Also, its name `daysTo` is not too generic to work with any enumeration type (unlike `next`).

But sure, it's possible to implement a function with the existing type signatures and call it, let's say, `elementsTo` or something like that 🙂

File: src/Lecture3.hs [hunk 3]
```diff
@@ -89,8 +95,23 @@ weekday to the second.
...
+    | from1 <= from2 = from2 - from1
+    | otherwise = abs (from2 - minB) + abs (from1 - maxB) + 1
+    where from1 = fromEnum el1
+          from2 = fromEnum el2
+          maxB = fromEnum (maxBound :: a)
+          minB = fromEnum (minBound :: a)
+
+-- This version was just fun to write
+daysTo1 :: (Eq a, Bounded a, Enum a) => a -> a -> Int
+daysTo1 d1 d2 = length $ takeWhile (/= d2) (iterate next d1)
```
Comment (chshersh) [inline src/Lecture3.hs:114]: Neat implementation 😅

File: src/Lecture3.hs [hunk 4]
```diff
@@ -178,7 +209,8 @@ together only different elements.
 Product {getProduct = 6}
 
 -}
-appendDiff3 = error "TODO"
+appendDiff3 :: (Monoid a, Eq a) => a -> a -> a -> a
+appendDiff3 a b c = mconcat $ nub [a, b, c]
```
Comment (chshersh) [inline src/Lecture3.hs]: That's a clever and concise implementation! 👏🏻 

However, the type signature is more strict than required. Your function works with `Monoid`. But not every type can have a `Monoid` instance. This means, that `appendDiff3` can't work with types like `List1`. But if you change the constraint from `Monoid` to `Semigroup`, this function will now work. Though, relaxing the constraint will require changing the implementation as well 🚧

**Review by chshersh — APPROVED** [review]
Wow, exercises to Lecture 3 are done! 🏆 
You've almost finished the course 🤯 

Keep up the great work and I hope you're enjoying the course so far 🤗 
The finish line is near 🏁

File: src/Lecture3.hs (hunk 4 above)
Comment (fetsh) [reply src/Lecture3.hs]: So something like this?
```haskell
appendDiff3 :: (Semigroup a, Eq a) => a -> a -> a -> a
appendDiff3 a b c = foldr1 (<>) $ nub [a, b, c]
```

Comment (chshersh) [reply src/Lecture3.hs]: That's one possible solution 👍🏻

Comment (fetsh) [reply src/Lecture3.hs]: Sounds like you've had a different expected/right solution :)

Btw: I've been thinking about a way to solve it without using nub, but without directly comparing all of the arguments also, and thought about the state monad. Would this function be a good one for learning/understanding the state monad? With something like {last appended :: a, setOfAllAppended :: Set a} as a state?

Comment (chshersh) [reply src/Lecture3.hs]: I was thinking about comparing relevant arguments with something like guards and appending only different elements 🙂 
This is good enough for for three arguments 👍🏻 

If you want to generalise this function to something like "list of elements" then an approach with a Set of already appended values works great 🔝  You don't need to use State monad for this though, it'll be a simple recursive function.

Comment (fetsh) [reply src/Lecture3.hs]: Yeah, I know that State monad is unnecessary in this case. It's just I've been trying to understand State monad and looking for a simplest case where can I apply it. Almost all the examples in "tutorials" are somehow complicated in one way or another. 
Anyway, I think you're right, and it'd be better to find a case where State monad is more appropriate. Thanks 👍

**Pull request merged**